Added setpoint generic flight controller function

bitcurious opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->
 
---
 
## Basic Info
 
| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | |
| Primary OS tested on | Ubuntu |
| Drone platform tested on | M300  |
 
---
 
## Description of contribution in a few bullet points
 
<!--I added this neat new feature -->
I've added setpoint generic functionality, allowing users to customize joystick control with a single control flag.
<!-- Also fixed a typo here -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
It will help user to have one topic to  control drone in different joystick mode using control flag.

 ## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Description of documentation updates required from your changes
 
<!-- Added new parameter, so need to add that to default configs and documentation page -->
<!-- I added some capabilities, need to document them -->

---
 
## Future work that may be required in bullet points
 
<!-- I think there might be some optimizations to be made -->
<!-- I see a lot of redundancy in this package -->

## Screenshots (if appropriate):

bitcurious committed:
Added setpoint generic flight controller function

bitcurious commented:
@biancabnd @vicmassy I’ve created a ROS 2 package with a script for testing. You can find the  here: [DJI RC Controller](https://github.com/bitcurious/dji_rc_controller.git). You can test the PR using this repository. It captures data from the DJI remote controller and publishes it to the setpoint generic topic with a control flag. Feel free to make any necessary changes if required.

biancabnd commented:
Hi @bitcurious, 
Thank you for your contribution and apologies for the delay; I've had limited time for PR reviews lately. I will try to come back to this and review it as soon as possible.

bitcurious commented:
@biancabnd No problem, take your time, also let me know if you want any additional control in [User input for controller](https://github.com/bitcurious/dji_rc_controller/blob/main/src/rc_controller.cpp#L102) like position or any different, or feel free to add that.

bitcurious committed:
Merge branch 'umdlife:main' into main

biancabnd commented on psdk_wrapper/src/modules/flight_control.cpp:
```diff
@@ -903,10 +903,236 @@ void
     const sensor_msgs::msg::Joy::SharedPtr msg)
 {
-  /** @todo implemnent generic control functionality */
+  /** @author Umesh Mane
+   * Contact: umeshmane280@gmail.com
+   */
+
```
Generic comments on this method. I think it is very long and makes it a bit hard to read. My suggestion would be to split it in several different methods (e.g. one method to set the horizontal axis, one for vertical, another for yaw control).

biancabnd commented on psdk_wrapper/src/modules/flight_control.cpp:
```diff
@@ -903,10 +903,236 @@ void
+  float x_cmd, y_cmd, z_cmd, yaw_cmd;
+  if (FRAME == Control::FRAME_GROUND)
+  {
+    FG = true;
```
do we really need these additional variables FG, FB? Couldn't you already set the joystick mode directly here? Or is there something I am missing

biancabnd commented on psdk_wrapper/src/modules/flight_control.cpp:
```diff
@@ -903,10 +903,236 @@ void
+    {
+      x_cmd = y_setpoint;
+      y_cmd = x_setpoint;
+      if (HORI == Control::HORIZONTAL_VELOCITY)
+      {
+        HV = true;
+      }
+      if (HORI == Control::HORIZONTAL_POSITION)
+      {
+        HP = true;
+      }
```
same comment here. Let's see if we can remove these additional variables HV, HP

biancabnd reviewed (commented):
Very nice addition @bitcurious!
I left you some comments to optimize a bit the method. Let me know what you think!

bitcurious committed:
updated setpoint genric function with different method

bitcurious commented:
@biancabnd I've updated the setpoint_generic function to make it more modular and easier to read as per your suggestions. I've also added logs to verify if the correct control mode is being selected; these can be removed after testing. The FRAME, HORI, and VERT are defined within each method to get updated values, let me know if I should leave them as is or make them global variables. Additionally, I added another file in the [dji_rc_controller](https://github.com/bitcurious/dji_rc_controller) package for selecting custom control modes as needed.

bitcurious committed:
updated brief of fn
